python/datashaper/datashaper/types/verb_callbacks.py

```diff
@@ -0,0 +1,56 @@
+from typing import Any, Protocol
+
+from datashaper.progress.types import Progress
+from datashaper.types.workflow_callbacks import WorkflowCallbacks
+
+
+class VerbCallbacks(Protocol):
+    """Provides a way to report status updates from the pipeline."""
+
+    def progress(self, progress: Progress):
+        "Report a progress update from the verb execution"
+        ...
+
+    def error(self, message: str, details: dict[str, Any] | None = None):
+        "Report a error from the verb execution."
+        ...
+
+    def warning(self, message: str, details: dict[str, Any] | None = None):
+        "Report a warning from verb execution."
+        ...
+
+    def log(self, message: str, details: dict[str, Any] | None = None):
+        "Report an informational message from the verb execution."
+        ...
+
+    def measure(self, name: str, value: float):
+        "Report a telemetry measurement from the verb execution."
+        ...
+
+
+class DelegatingVerbCallbacks(VerbCallbacks):
```

Comment:
With Protocols you don't need inheritance, the interpreter should be able to detect the proper protocol implementation.

Reply:
Right, so my thought here was that if a user was only interested in implementing a couple of workflow hooks, then they could subclass the base class and then define the specific functions they want. We could provide a protocol and a base class, but that seems like overkill to me

Reply:
I went ahead and reverted to a protocol, but I added a new delegating implementation that guarantees all the methods are available.